asmcomp/cmm_helpers.ml

```diff
@@ -2215,29 +2212,28 @@ type binary_primitive = expression -> expression -> Debuginfo.t -> expression
 
 (* Helper for compilation of initialization and assignment operations *)
 
-type assignment_kind = Caml_modify | Caml_initialize | Simple
+type assignment_kind = Caml_modify | Simple
 
 let assignment_kind
     (ptr: Lambda.immediate_or_pointer)
     (init: Lambda.initialization_or_assignment) =
   match init, ptr with
   | Assignment, Pointer -> Caml_modify
-  | Heap_initialization, Pointer -> Caml_initialize
+  | Heap_initialization, _ ->
+     Misc.fatal_error "Cmm_helpers: Lambda.Heap_initialization unsupported"
   | Assignment, Immediate
-  | Heap_initialization, Immediate
   | Root_initialization, (Immediate | Pointer) -> Simple
 
 let setfield n ptr init arg1 arg2 dbg =
   match assignment_kind ptr init with
   | Caml_modify ->
+      (* In principle, we could track whether the block being modified
+         is possibly local and use caml_modify when it's not. However,
+         the extra complexity only saves a single branch, so for now
+         this unconditionally uses caml_modify_local *)
```

Comment:
Possibly we should add this tracking now. I'm not that worried about the branch, but I think that people have various tricks that try to track `caml_modify` and so I'd rather we still used it in code that isn't using local allocations.

Reply:
When I wrote the previous comment, I hadn't yet read the implementation of `caml_modify_local`. I assumed it would have the body of `caml_modify` hand-inlined into it. Since it doesn't I think that people's tricks for tracking `caml_modify` will probably still work.